ipam/ipam.go

```diff
@@ -0,0 +1,250 @@
+
+	CommonDomain    string
+	HostClusterName string
+}
+
+type IPAM struct {
+	hostFramework *framework.Host
+	logger        micrologger.Logger
+	provider      provider.Interface
+
+	commonDomain    string
+	hostClusterName string
+}
+
+func New(config Config) (*IPAM, error) {
+	if config.HostFramework == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostFramework must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+	if config.Provider == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Provider must not be empty", config)
+	}
+	if config.CommonDomain == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CommonDomain must not be empty", config)
+	}
+	if config.HostClusterName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.HostClusterName must not be empty", config)
+	}
+
+	s := &IPAM{
+		hostFramework: config.HostFramework,
+		logger:        config.Logger,
+		provider:      config.Provider,
+
+		commonDomain:    config.CommonDomain,
+		hostClusterName: config.HostClusterName,
+	}
+
+	return s, nil
+}
+
+func (c *IPAM) Test(ctx context.Context) error {
+
+	var (
+		// Clusters to be created during this test.
+		clusterOne   = c.hostClusterName + "-cluster0"
+		clusterTwo   = c.hostClusterName + "-cluster1"
+		clusterThree = c.hostClusterName + "-cluster2"
+		clusterFour  = c.hostClusterName + "-cluster3"
+
+		// allocatedSubnets[clusterName]subnetCIDRStr
+		allocatedSubnets = make(map[string]string)
+		err              error
+		// guestFrameworks[clusterName]guestFramework
+		guestFrameworks = make(map[string]*framework.Guest)
+	)
+
+	defer func() {
+		for _, cn := range []string{clusterOne, clusterTwo, clusterThree, clusterFour} {
+			err := c.provider.DeleteCluster(cn)
+			if err != nil {
+				c.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("cluster %s deletion failed: %#v", cn, err))
+			}
+		}
+	}()
+
+	// This is a list of clusters that are created in first test phase.
+	clusters := []string{clusterOne, clusterTwo, clusterThree}
+
+	{
+
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating three guest clusters: %#v", clusters))
+
+		for _, cn := range clusters {
+			err = c.provider.CreateCluster(cn)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+			c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created guest cluster %s", cn))
+		}
+	}
+
+	{
+		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("waiting for three guest clusters to be ready: %#v", clusters))
+
+		for _, cn := range clusters {
+			cfg := framework.GuestConfig{
+				Logger: c.logger,
+
+				ClusterID:    cn,
+				CommonDomain: c.commonDomain,
+			}
+			guestFramework, err := framework.NewGuest(cfg)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			guestFrameworks[cn] = guestFramework
+			err = guestFramework.Setup()
```

Comment:
Does it make sense to run it in parallel?

Reply:
I don't think it matters here. The above loop with calls to `CreateCluster()` creates `AWSConfigs` which triggers cluster creation in parallel. This is only checking that API is up. It would need to wait for the last one anyway so I don't see major gain in speed but it would be a bit more complicated to read.